core/server/data/migrations/versions/5.0/2022-03-28-15-25-backfill-mrr-adjustments-for-offers.js

```diff
@@ -0,0 +1,109 @@
+            .select('*')
+            .from('members_paid_subscription_events')
+            .where('type', 'created')
+            .whereIn('member_id', memberIds);
+
+        function storeEventOnMemberId(storage, event) {
+            return {
+                ...storage,
+                [event.member_id]: storage[event.member_id] ? storage[event.member_id].concat(event) : [event]
+            };
+        }
+
+        const mrrCreatedEventsByMemberId = mrrCreatedEvents.reduce(storeEventOnMemberId, {});
+
+        const updatedEvents = [];
+
+        function calculateMRR(subscription, redemption) {
+            if (redemption && subscription.interval !== redemption.discount_interval) {
+                logging.error('Found invalid price & redemption pair');
+                logging.info(redemption.subscription_id);
+                return calculateMRR(subscription);
+            }
+
+            if (!redemption) {
+                return subscription.interval === 'year' ? subscription.amount / 12 : subscription.amount;
+            }
+
+            if (redemption.discount_type === 'percent') {
+                return calculateMRR({
+                    interval: subscription.interval,
+                    amount: subscription.amount * (100 - redemption.discount_amount) / 100
+                });
+            }
+
+            return calculateMRR({
+                interval: subscription.interval,
+                amount: subscription.amount - redemption.discount_amount
+            });
+        }
+
+        offerRedemptions.forEach((redemption) => {
+            const memberEvents = mrrCreatedEventsByMemberId[redemption.member_id];
+
+            // If a member has had multiple subscriptions we ignore because we cannot easily work out which event is correct.
+            if (memberEvents.length !== 1) {
+                return;
+            }
+
+            const firstEvent = memberEvents[0];
+
+            const mrr = calculateMRR({
+                interval: redemption.interval,
+                amount: redemption.amount
+            }, redemption);
+            logging.info(mrr);
+            updatedEvents.push({
+                id: firstEvent.id,
+                type: 'created',
+                subscription_id: redemption.subscription_id,
+                member_id: firstEvent.member_id,
+                from_plan: firstEvent.from_plan,
+                to_plan: firstEvent.to_plan,
+                currency: firstEvent.currency,
+                source: firstEvent.source,
+                created_at: firstEvent.created_at,
+                mrr_delta: mrr
+            });
+        });
+
+        if (updatedEvents.length === 0) {
+            return;
+        }
+
+        const idsToDelete = updatedEvents.map(event => event.id);
+
+        await knex('members_paid_subscription_events').whereIn('id', idsToDelete).del();
+        await knex('members_paid_subscription_events').insert(updatedEvents);
```

Comment:
As previously mentioned, I think we should `.batchInsert` here

Reply:
Yup!